src/rust/engine/process_execution/docker/src/docker.rs

```diff
@@ -393,147 +397,131 @@ impl process_execution::CommandRunner for CommandRunner<'_> {
                         exclusive_spawn,
-                    )
-                    .map_err(|msg| {
-                        // Processes that experience no infrastructure issues should result in an "Ok" return,
-                        // potentially with an exit code that indicates that they failed (with more information
-                        // on stderr). Actually failing at this level indicates a failure to start or otherwise
-                        // interact with the process, which would generally be an infrastructure or implementation
-                        // error (something missing from the sandbox, incorrect permissions, etc).
-                        //
-                        // Given that this is expected to be rare, we dump the entire process definition in the
-                        // error.
-                        ProcessError::Unclassified(format!(
-                            "Failed to execute: {req_debug_repr}\n\n{msg}"
-                        ))
-                    })
-                    .await;
-
-                match &res {
-                    Ok(_) => workunit.increment_counter(Metric::DockerExecutionSuccesses, 1),
-                    Err(_) => workunit.increment_counter(Metric::DockerExecutionErrors, 1),
+                    ).await;
+                    match &res {
+                        Ok(_) => workunit.increment_counter(Metric::DockerExecutionSuccesses, 1),
+                        Err(_) => workunit.increment_counter(Metric::DockerExecutionErrors, 1),
+                    };
+                    if keep_sandboxes == KeepSandboxes::Always || (
+                        keep_sandboxes == KeepSandboxes::OnFailure &&
+                            res.as_ref().map(|r| r.exit_code).unwrap_or(1) != 0
+                    ) {
+                        workdir.keep(&mreq.description);
+                        setup_run_sh_script(
+                            workdir.path(),
+                            &mreq.env,
+                            &mreq.working_directory,
+                            &mreq.argv,
+                            workdir.path(),
+                        )?;
+                    }
+                    res
                 }
-
-                if keep_sandboxes == KeepSandboxes::Always
-                    || keep_sandboxes == KeepSandboxes::OnFailure
-                        && res.as_ref().map(|r| r.exit_code).unwrap_or(1) != 0
-                {
-                    workdir.keep(&req.description);
-                    setup_run_sh_script(
-                        workdir.path(),
-                        &req.env,
-                        &req.working_directory,
-                        &req.argv,
-                        workdir.path(),
-                    )?;
+            ).await;
+            match process_result {
+                Err(CapturedWorkdirError::Retryable { status, message }) => {
+                    errors.push(message);
+                    if status == 404 {
```

Comment:
We check for the 404 constant in several places. Admittedly many people will instantly recognise that as likely being some sort of not-found error code, but still... could we pull it out to a `const`?